### PR TITLE
Wishlist Service

### PR DESCRIPTION
**Pull request by legendofa**

### CODE REVIEW

File: compose.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+services:
+  app:
```
Comment (nk-coding) [inline compose.yaml:2]: this one currently does not work yet

File: src/mutation_root.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
+use std::collections::HashSet;
+
+use async_graphql::{Object, Context, FieldResult, Error};
+use mongodb::{Collection, bson::{doc, DateTime}};
+use uuid::Uuid;
+
+use crate::wishlist::Wishlist;
+pub struct MutationRoot;
+
+fn uuid_parse(id: &String) -> Result<String, Error> {
```
Comment (nk-coding) [inline src/mutation_root.rs]: missing documentation

**Review by nk-coding — CHANGES_REQUESTED** [review]
I can't say too much about the code due to missing rust knowledge, maybe we have a team member with knowledge there. To me it looks good, except missing code documentation.

However I have a few notes about the GraphQL schema.
Note that I don't think all of this has to be done as part of this PR, but it would be good to create issues in Gropius for the stuff you decide to do in future PRs.
- most or all fields are missing documentation
- id inputs and outputs should use the `ID` scalar as type, not `String`
- mutations usually only take one input parameter named after the mutation (e.g. `AddWishlistInput` for `addWishlist`, which then contains all fields. IMHO this is a good style guide to follow here
- add/update mutations should return the created/updated object
- a many query should provide pagination features via the adapted connection standard we decide on. Here, in particular ordering and limiting are missing.
- As part of the GraphQL federation standard, one should not return the id of a related entity, but instead the entity itself (in this service the entities User and ProductVariant will then only consist of an id)
- `createdAt` / `lastUpdatedAt` should use the `DateTime` scalar
  definition from catalog: 
  ```graphql
  "A slightly refined version of RFC-3339 compliant DateTime Scalar"
  scalar DateTime
  ```
- `productVariantIds` should be also a connection with pagination support

**Comment by legendofa** [comment]
Shall i also wrap mutations with only one argument, as delete, in a `DeleteWishlistInput` object?

**Comment by nk-coding** [comment]
> Shall i also wrap mutations with only one argument, as delete, in a `DeleteWishlistInput` object?

Delete is a hard one. I do not really have a fixed opinion about this.  For create and update with one input, I would still say yes. Regarding delete, a benefit would be consistency, however I do also see why one would argue this is completely redundant.
One argument could be that in some cases, even delete could need more arguments, .e.g. a flag if some other entities should be deleted with the entity.
In any case, we should come to a decision we follow in all services.

**Comment by legendofa** [comment]
Why should created/update return the object, isn't that overhead? Also did not find out yet how to implement pagination

**Comment by legendofa** [comment]
> > Shall i also wrap mutations with only one argument, as delete, in a `DeleteWishlistInput` object?
> 
> Delete is a hard one. I do not really have a fixed opinion about this. For create and update with one input, I would still say yes. Regarding delete, a benefit would be consistency, however I do also see why one would argue this is completely redundant. One argument could be that in some cases, even delete could need more arguments, .e.g. a flag if some other entities should be deleted with the entity. In any case, we should come to a decision we follow in all services.

:+1:  I just left it concise for now, we can decide this in the next meeting then :)

**Comment by nk-coding** [comment]
> Why should created/update return the object, isn't that overhead? Also did not find out yet how to implement pagination

First, for create, if you don't return it, how should the frontend know the id of the created object?
Second, often, when you create or update sth, you want to display the created/updated thing. Returning it allows you to do so without having to make another query.
Also, regarding efficiency, there should not be too much overhead. Returning it allows the user to fetch information from the created/update entity, however, GraphQL does not require you to do so (except at least one scalar field, but that is minimal overhead imho).

**Comment by legendofa** [comment]
> First, for create, if you don't return it, how should the frontend know the id of the created object? Second, often, when you create or update sth, you want to display the created/updated thing. Returning it allows you to do so without having to make another query. Also, regarding efficiency, there should not be too much overhead. Returning it allows the user to fetch information from the created/update entity, however, GraphQL does not require you to do so (except at least one scalar field, but that is minimal overhead imho).

Sounds good, i have adapted it to return the UUID on creation and modification now :)

**Comment by nk-coding** [comment]
> > First, for create, if you don't return it, how should the frontend know the id of the created object? Second, often, when you create or update sth, you want to display the created/updated thing. Returning it allows you to do so without having to make another query. Also, regarding efficiency, there should not be too much overhead. Returning it allows the user to fetch information from the created/update entity, however, GraphQL does not require you to do so (except at least one scalar field, but that is minimal overhead imho).
> 
> Sounds good, i have adapted it to return the UUID on creation and modification now :)

Again, it would be better to return the whole object to allow the user to decide what to get. In GraphQL, you basically never return just an id.
We should probably discuss if we want to support the Payload pattern (wrap the output in a Payload object).
See https://blog.logrocket.com/anti-patterns-graphql-schema-design/ (insufficient mutation response)

**Review by nk-coding — CHANGES_REQUESTED** [review]
- WishlistOrder should have fields `direction` and `field` to be consistent with other servics
- Missing documentation for WishlistOrder, WishlistOrderField, OrderDirection
- QueryRoot should be named Query (federation support)
- MutationRoot should be named Mutation (federation support)
- if possible, use ID scalar instead of custom UUID scalar (maybe just rename)
- wishlists query should return a WishlistConnection which also allows to get totalCount, hasNextPage
- field inputs are often not documented, instead the documentation is part of the field itself (however there is also documentation for ctx which does not exist in GraphQL)

**Comment by legendofa** [comment]
Would it be possible to use UUIDs directly, as IDs lead to more error handling stuff? We only use UUIDs so no general IDs should be required

**Comment by legendofa** [comment]
>     * field inputs are often not documented, instead the documentation is part of the field itself (however there is also documentation for ctx which does not exist in GraphQL)

Removed the `ctx`, do we really need duplicate docs there? The inputs are already the documented input object.

**Comment by nk-coding** [comment]
> >     * field inputs are often not documented, instead the documentation is part of the field itself (however there is also documentation for ctx which does not exist in GraphQL)
> 
> Removed the `ctx`, do we really need duplicate docs there? The inputs are already the documented input object.
> 

Well if I could choose, I would add it only to the inputs themselves.

**Comment by legendofa** [comment]
I have added Connection support in this PR. Changed the field docs to the annotation style, but i am not sure if it turns out the way we want it to be. Next up will be the federation support, but will create a new PR for that.

**Review by nk-coding — CHANGES_REQUESTED** [review]
- In GraphQL, maybe rename BaseConnection to WishlistConnection as it is a connection for Wishlists (also we will have clashes otherwise in the future).
- BaseConnection and Wishlist are both missing GraphQL descriptions for all fields

**Comment by legendofa** [comment]
> * In GraphQL, maybe rename BaseConnection to WishlistConnection as it is a connection for Wishlists (also we will have clashes otherwise in the future).
> 
>     * BaseConnection and Wishlist are both missing GraphQL descriptions for all fields

I am aware of this and did not find out how to rename it, there is no such thing as inheritance in Rust, which would normally take care of this. I could make it non-generic instead.

**Review by nk-coding — APPROVED** [review]

**Pull request merged**